File: Octave/gradiente_conjugado.m
## Copyright (C) 2019 mherr
## 
## This program is free software: you can redistribute it and/or modify it
## under the terms of the GNU General Public License as published by
## the Free Software Foundation, either version 3 of the License, or
## (at your option) any later version.
## 
## This program is distributed in the hope that it will be useful, but
## WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
## 
## You should have received a copy of the GNU General Public License
## along with this program.  If not, see
## <https://www.gnu.org/licenses/>.

## -*- texinfo -*- 
## @deftypefn {} {@var{retval} =} descenso_coordinado (@var{input1}, @var{input2})
##
## @seealso{}
## @end deftypefn

## Author: mherr <mherr@MSI>
## Created: 2019-08-27

## Metodo iterativo del Gradiente Conjugado para optimizacion en varias
## variables.
## Entradas: valores iniciales, variables de la funcion, tolerancia minima
## del resultado, funcion a evaluar en string y regla a utilizar.
## Salidas:Valor aproximado y numero de iteraciones
function [x_k, iteration] = gradiente_conjugado(x0, str_variables, tol, funcion, regla_bk = 'FR')
    n = length(str_variables);
    f = str2func(funcion);
    derivatives = sym(zeros(1, n));
    for i = 1:n
        variables(i) = sym(str_variables(i));
    endfor
    
    for i = 1:n
        derivatives(i) = diff(f, variables(i));
        i += 1;
    endfor
    x_k = x0;
    error = tol;
    iteration = 0;
    
    function [grad] = gradiente(derivatives, variables, values, n)
        grad = zeros(1, n);
        
        for i = 1:n
            grad(i) = matlabFunction(derivatives(i))(num2cell(values){:});
            i += 1;
        endfor
    endfunction
    
    g_k_1 = g_k = gradiente(derivatives, variables, x_k, n);
    d_k = -1 * g_k;
    
    function [alpha] = get_alpha(f, variables, x_k, d_k, g_k)
        alpha = 1;
        delta = rand(1);
        f_arg = x_k + alpha * d_k;
        while double(f(num2cell(f_arg){:})) - double(f(num2cell(x_k){:})) > delta * alpha * g_k * transpose(d_k)
            alpha /= 2;
            f_arg = x_k + alpha * d_k;
        endwhile    
    endfunction
   
    function [beta] = get_beta(regla, g_k_1, g_k, d_k)
        if lower(regla) == 'cd'
            beta = norm(g_k_1)**2 / ((-1 * d_k)*transpose(g_k));
        elseif lower(regla) == 'dy'
            beta = norm(g_k_1)**2 / d_k*transpose(g_k_1 - g_k);
        else
            beta = norm(g_k_1)**2 / norm(g_k)**2;
        end    
    endfunction
    
    while error >= tol
        alpha = get_alpha(f, variables, x_k, d_k, g_k_1);
        x_k = x_k + alpha * d_k;
        g_k = g_k_1;
        g_k_1 = gradiente(derivatives, variables, x_k, n);
        d_k = -1 * g_k_1 + get_beta(regla_bk, g_k_1, g_k, d_k) * d_k;
        error = norm(g_k_1);
        iteration += 1;
    endwhile
endfunction

File: Python/gradiente_conjugado.py
from sympy import *
import numpy as np
from numpy import linalg
import random


def gradiente_conjugado(x0, variables, tol, funcion, regla_bk='FR'):
    variables = symbols(variables)
    funcion = sympify(funcion)
    f = lambdify(variables, funcion, "numpy")
    derivatives = []
    for variable in variables:
        derivatives.append(lambdify(variables, diff(funcion, variable)))
    x_k = np.array([x0])
    g_k_1 = g_k = gradiente(derivatives, x_k)
    d_k = -1 * g_k
    iteracion = 0
    error = tol
    while error >= tol:
        alpha = get_alpha(f, x_k, d_k, g_k_1, 1)
        x_k = x_k + alpha * d_k
        g_k = g_k_1
        g_k_1 = gradiente(derivatives, x_k)
        d_k = -1 * g_k_1 + beta(regla_bk, g_k_1, g_k, d_k) * d_k
        error = linalg.norm(g_k_1)
        iteracion += 1
    return x_k[0].tolist(), iteracion


def gradiente(derivatives, values):
    result = []
    for derivative in derivatives:
        result.append(derivative(*values[0]))
    return np.array([result])


def get_alpha(f, x_k, d_k, g_k, alpha):
    delta = random.uniform(0, 1) #0.5 #random.randint(1, 9) * 0.1
    f_arg = x_k + alpha * d_k
    while f(*f_arg[0]) - f(*x_k[0]) > delta * alpha * g_k.dot(d_k.T):
        alpha /= 2
        f_arg = x_k + alpha * d_k
    return alpha


def beta(regla, g_k_1, g_k, d_k):
    if regla.lower() == 'cd':
        b = linalg.norm(g_k_1)**2 / ((-1 * d_k).dot(g_k.T))
    elif regla.lower() == 'dy':
        b = linalg.norm(g_k_1)**2 / d_k.dot((g_k_1 - g_k).T)
    else:
        b = linalg.norm(g_k_1)**2 / linalg.norm(g_k)**2
    return b


print(gradiente_conjugado([0, 3], ['x', 'y'], 0.001, '(x-2)**4+(x-2*y)**2', 'fr'))
print(gradiente_conjugado([0, 3], ['x', 'y'], 0.001, '(x-2)**4+(x-2*y)**2', 'cd'))
print(gradiente_conjugado([0, 3], ['x', 'y'], 0.001, '(x-2)**4+(x-2*y)**2', 'dy'))

print(gradiente_conjugado([1, 1, 1], ['x', 'y', 'z'], 0.001, '(x-2)**2+ (y+3)**2 + (x + y + z)**2', 'fr'))
print(gradiente_conjugado([1, 1, 1], ['x', 'y', 'z'], 0.001, '(x-2)**2+ (y+3)**2 + (x + y + z)**2', 'cd'))
print(gradiente_conjugado([1, 1, 1], ['x', 'y', 'z'], 0.001, '(x-2)**2+ (y+3)**2 + (x + y + z)**2', 'dy'))
